src/keywords/ref_.rs

```diff
@@ -1,60 +1,74 @@
 use super::{CompilationResult, Validate};
 use crate::{
     compilation::{compile_validators, CompilationContext, JSONSchema},
-    error::{error, ErrorIterator},
+    error::{error, ErrorIterator, ValidationError},
+    keywords::Validators,
 };
 use serde_json::Value;
 use url::Url;
+use parking_lot::RwLock;
 
 pub struct RefValidator {
     reference: Url,
+    /// Precopmuted validators.
+    /// They are behind a RwLock as is not possible to compute them
+    /// at compile time without risking infinite loops of references
+    /// and at the same time during validation we iterate over shared
+    /// references (&self) and not owned references (&mut self).
+    validators: RwLock<Option<Validators>>,
 }
 
 impl RefValidator {
     #[inline]
-    pub(crate) fn compile(reference: &str, context: &CompilationContext) -> CompilationResult {
+    pub(crate) fn compile(
+        reference: &str,
+        context: &CompilationContext,
+    ) -> CompilationResult {
         let reference = context.build_url(reference)?;
-        Ok(Box::new(RefValidator { reference }))
+        Ok(Box::new(RefValidator {
+            reference,
+            validators: RwLock::new(None),
+        }))
+    }
+
+    /// Ensure that validators are built and built once.
+    fn ensure_validators<'a>(&self, schema: &'a JSONSchema) -> Result<(), ValidationError<'a>> {
+        if self.validators.read().is_none() {
+            let (scope, resolved) = schema
+                .resolver
+                .resolve_fragment(schema.draft, &self.reference, schema.schema)?;
+            let context = CompilationContext::new(scope, schema.draft);
+            let validators = compile_validators(&resolved, &context)?;
+            
+            // Inject the validators into self.validators
+            *self.validators.write() = Some(validators);
+        }
+        Ok(())
     }
 }
 
 impl Validate for RefValidator {
     fn validate<'a>(&self, schema: &'a JSONSchema, instance: &'a Value) -> ErrorIterator<'a> {
-        match schema
-            .resolver
-            .resolve_fragment(schema.draft, &self.reference, schema.schema)
-        {
-            Ok((scope, resolved)) => {
-                let context = CompilationContext::new(scope, schema.draft);
-                match compile_validators(&resolved, &context) {
-                    Ok(validators) => Box::new(
-                        validators
-                            .into_iter()
-                            .flat_map(move |validator| validator.validate(schema, instance)),
-                    ),
-                    Err(e) => error(e.into()),
-                }
-            }
-            Err(e) => error(e),
+        if let Err(err) = self.ensure_validators(schema) {
+            error(err)
+        } else {
+            Box::new(
+                self.validators.read().as_ref().expect("ensure_validators guarantees the presence of the validators")
```

Comment:
With having that I suspect that it might improve our current situation with cloning - now validators are moved from this scope to the `RefValidator` which means that we will not return anything owned by the current function and we can try to return references instead. What do you think? I am not completely sure about lifetimes

Reply:
The `RefValidator::validate` method needs to `collect` the `FlatMap` as `self.validators` is owned by the `RwLockGuard`. This has an impact on materialising a new `Vec` to contain the errors, but we're not cloning duplicating `ValidationError` instances.

I don't think that we might have ways around this (except if we decide to inject `unsafe`) which I would prefer not doing as performance improvements are already good enough.

We might, in the future, try to inspect all the cases where we have `...collect().into_iter()` an `ValidationError::into_owned` calls to try to remove them as they are indicators of creating memory just to make rust borrow-checker happy ;)

Reply:
:+1: